### DISCUSSION TITLE
pySPlisHSPlasH uses wrong directory for output, data and resources

### DISCUSSION THREAD

**Discussion by lukashof** [InteractiveComputerGraphics/SPlisHSPlasH#283] [Q&A] [answered]
Hi 👋, 

I am running pySPlisHSPlasH on Windows inside a **venv**. The installation seems to have worked. I do have the `data` and `resources` folders inside my venv. However, when I run the examples or `splash`, the output directory is set to my **global** Python installation and SPlisHSPlasH tries to load shaders or models from there as well and fails as the folders are missing.

I can of course set the output directory manually. But the only workaround I have found for the "missing" shaders and models, is to copy `data` and `resources` into my global python installation.

Did I make a mistake during the installation? I pointed to the Python executable inside the venv when compiling and built a wheel then ran the installation of pySPlisHSPlasH from inside my venv. Do I have to specify the executable again at this point?

This problem seems similar to #254 and the discussion under #151.

Best regards!

**Comment by digitalillusions** [comment]
Thanks for letting us know. I have not been aware of this issue actually, since I always use anaconda virtual environments, where this issue does not occur. I have also confirmed the behavior that you describe when using `virtualenv`.

It seems that the cause of the problem is the fact that `virtualenv` creates a symlink to the global python executable. Splishsplash gets the executable path here https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/blob/b7f8321bf999fed3bcd31d00a6070a23458e385a/Utilities/FileSystem.h#L215-L229 which apparently resolves through the symlink to the original executable and then tries to load all assets relative to this path. I'm not exactly sure how to prevent this behavior, since this is also tied to how the OS handles executables. I'll update this discussion if I find a solution.

**Reply by digitalillusions** [reply]
Upon closer inspection, you can use the `--copies` [option](https://docs.python.org/3/library/venv.html#creating-virtual-environments) when creating your virtual environment to make a copy of the python executable as follows.

```sh
python -m venv --copies venv-name
```

In my tests this has solved the issue of loading assets correctly without having to copy them into the global python installation. I hope this helps.

**Reply by lukashof** [reply]
Thanks so much for taking the time to test this out! I am sorry it took me so long to get back to this!

I tried to use  `--copies`. Unfortunately, I got the same result as previously and the assets were missing in the venv. Even on another machine, I've had no luck with this.

I tried using Conda but got a compiler error. I'm [including](https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/files/13640677/compiler_error_conda.txt) the error message, but this might be unrelated to the problem.

I also can't tell whether the `--copies` option did anything at all. The venv-Folder is the same size and the `pyvenv.cfg` looks exactly the same.

This really is a minor issue on my personal machine, but I think I will annoy other users of our servers if I have to mess with the global interpreter on those machines 😬.

**Reply by lukashof** [reply]
Sorry to bother you about this again, but do you have another idea regarding this issue? I tried to find out where the wrong path is coming from but couldn't find it.

**Reply by digitalillusions** [reply]
Honestly I would love to help, but this is not anything I've encountered on any platform. The `--copies` option works for me when using `venv`, although this was on a linux system. I unfortunately don't have access to a windows machine at the moment, but even there I had no trouble with anaconda virtual environments in the past. If you are working on a linux server, the anaconda virtual environment method should work without any issues. If you get errors on linux I might be able to help out more.

EDIT: The error message for anaconda also means nothing to me. I would try to delete the build folder and try running `python setup.py bdist_wheel` again, but this will only get you so far.

**Reply by lukashof** [reply]
I see, maybe it is a Windows-only problem then. So far, I haven't tried it on a Linux system. I guess I'll try that or with Conda again.

This might be obvious to you already, but I found that `GetModuleFileName` returns the path to my global interpreter. It is used here:

https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/blob/dc2f63f3188cfe5a6fac8f6837de1335782958be/Utilities/FileSystem.h#L215-L219

**Reply by digitalillusions** [reply]
It could be that I didn't catch the problem then, or that `--copies` doesn't actually copy the interpreter and necessary DLLs to the `venv` directory on windows. Again, I unfortunately cannot test myself this at the moment. Can you check if the `venv/bin` directory actually contains its own **non-shortcutted** version of python?

**Reply by lukashof** [reply]
The folder structure on Windows seems to be a bit different, so there is no `bin`. However, there is an executable inside the `Scripts` folder.
![grafik](https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/assets/100953722/e98a62a7-b6ca-4186-9d1a-53046e63b617)

I can see no difference running the command with or without the `--copies` option, but it will throw an error when run with `--symlinks`.

**Reply by lukashof** [reply]
I just realized, that `getProgramPath` was probably intended to return the path to `bin` containing the `SPHSimulator.exe`. Or am I mistaken? In that case it wouldn't be an issue with `venv` at all.

**Reply by lukashof** [reply]
I have now tried this on a Linux system and the resources are indeed picked up from inside the `venv`-folder.

Regarding Windows, it might be an option to get the path to the interpreter during building instead of using `GetModuleFileName` if the bindings are used. Then again, it might be more intuitive to have the resources and outputs somewhere in the project folder or in the build directory. That seems more intuitive to me. Or am I missing something?

**Comment by lukashof** [comment, marked as answer]
I tried solving this in #368.
